### PR TITLE
refactor(connector): Added amount conversion framework to Wise.

### LINKED ISSUE #6121: [REFACTOR]: [WISE] Add amount conversion framework to Wise
### :memo: Feature Description

Currently, amounts are represented as `i64` values throughout the application. We want to introduce a `Unit` struct that explicitly states the denomination. A new type, `MinorUnit`, has been added to standardize the flow of amounts across the application. This type will now be used by all the connector flows.

Rather than handling conversions in each connector, we will centralize the conversion logic in one place within the core of the application.

### :hammer: Possible Implementation

- For each connector, we need to create an amount conversion function. Connectors will specify the format they require, and the core framework will handle the conversion accordingly.
- Connectors should invoke the `convert` function to receive the amount in their required format.
- Refer to the [connector documentation](https://docs.wise.com/api-docs/api-reference/card) to determine the required amount format for each connector.
- You can refer [this PR](https://github.com/juspay/hyperswitch/pull/4825) for more context.

🔖 Note: All the changes needed should be contained within `hyperswitch/crates/router/src/connector/` , `crates/router/src/types/api.rs` , `crates/router/tests/connectors/`

### :package: Have you spent some time checking if this feature request has been raised before?

- [ ] I checked and didn't find a similar issue

### :package: Have you read the Contributing Guidelines?

- [ ] I have read the [Contributing Guidelines](https://github.com/juspay/hyperswitch/blob/main/docs/CONTRIBUTING.md)

### :package: Submission Process:
- Ask the maintainers for assignment of the issue, you can request for assignment by commenting on the issue itself.
- Once assigned, submit a pull request (PR).
- Maintainers will review and provide feedback, if any. 
- Maintainers can unassign issues due to inactivity, [read more here](https://github.com/juspay/hyperswitch/wiki/Hacktoberfest-Contribution-Rules).

Refer [here](https://github.com/juspay/hyperswitch/blob/main/docs/TERMS_OF_CONTEST.md) for Terms and conditions for the contest.

### PR DESCRIPTION
**Pull request by shuklamaneesh23**
## Type of Change

- [ ] Bugfix
- [ ] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

This pull request introduces significant changes to the Wise connector in the `crates/router` module, focusing on the integration of a new amount conversion utility and refactoring to utilize a new `WiseRouterData` struct. The changes enhance how amounts are handled and improve the overall structure and clarity of the code.

### Refactoring and Enhancements:

* **Amount Conversion Utility Integration:**
  - Added `amount_converter` to the `Wise` struct and integrated it into the `new` method.
  - Updated `convert_amount` usage in `WisePayoutQuoteRequest` and `WiseRecipientCreateRequest` implementations to use the new `amount_converter`. [[1]](diffhunk://#diff-a149ab1c35f89dd3b0eed2fca15df170708e81962550226b958ae6c367515043L365-R382) [[2]](diffhunk://#diff-a149ab1c35f89dd3b0eed2fca15df170708e81962550226b958ae6c367515043L444-R467)

* **Introduction of `WiseRouterData` Struct:**
  - Added `WiseRouterData` struct to encapsulate amount and router data, and implemented conversion from a tuple.
  - Modified `WisePayoutQuoteRequest` and `WiseRecipientCreateRequest` to use `WiseRouterData` for transformations. [[1]](diffhunk://#diff-1e46db110c015ea743fbbe850bf756011b4311a1ef56fe5ee75c911bb4960b15L351-R371) [[2]](diffhunk://#diff-1e46db110c015ea743fbbe850bf756011b4311a1ef56fe5ee75c911bb4960b15L423-R451)

### Codebase Simplification:

* **Removal of Unnecessary Debug Trait:**
  - Removed the `Debug` trait from the `Wise` struct.

* **Update in Connector Initialization:**
  - Updated the initialization of the `Wise` connector to use the `new` method in `api.rs` and test files. [[1]](diffhunk://#diff-71935350b305eb696195bc45f26e6f2a16097fb17bd5ac55192eb1d03558bf9aL483-R483) [[2]](diffhunk://#diff-0177b5efcc163c3f657a8c60b0162bf272a68a5e22beb13f6eb309bd34d54166L31-R31)

### Minor Enhancements:

* **Type Adjustments:**
  - Changed various fields in `WiseRecipientCreateRequest` and `WisePayoutQuoteRequest` to use `StringMajorUnit` instead of `String` or `i64`. [[1]](diffhunk://#diff-1e46db110c015ea743fbbe850bf756011b4311a1ef56fe5ee75c911bb4960b15L64-R79) [[2]](diffhunk://#diff-1e46db110c015ea743fbbe850bf756011b4311a1ef56fe5ee75c911bb4960b15L159-R175)

These changes collectively improve the robustness and maintainability of the Wise connector, ensuring better handling of amount conversions and a cleaner code structure.

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

Fixes #6121 
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: crates/router/src/connector/wise.rs [hunk 1]
```diff
@@ -27,8 +28,18 @@ use crate::{
...
-pub struct Wise;
+#[derive(Clone)]
+pub struct Wise {
+    amount_converter: &'static (dyn AmountConvertor<Output = StringMajorUnit> + Sync),
+}
+
+impl Wise {
+    pub fn new() -> &'static Self {
+        &Self {
+            amount_converter: &StringMajorUnitForConnector,
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise.rs:39]:
```suggestion
            amount_converter: &MinorUnitForConnector,
```

File: crates/router/src/connector/wise.rs [hunk 2]
```diff
@@ -27,8 +28,18 @@ use crate::{
     utils::BytesExt,
 };
 
-#[derive(Debug, Clone)]
-pub struct Wise;
+#[derive(Clone)]
+pub struct Wise {
+    amount_converter: &'static (dyn AmountConvertor<Output = StringMajorUnit> + Sync),
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise.rs:33]:
```suggestion
    amount_converter: &'static (dyn AmountConvertor<Output = MinorUnit> + Sync),
```

File: crates/router/src/connector/wise.rs [hunk 3]
```diff
@@ -1,15 +1,16 @@
 pub mod transformers;
-use std::fmt::Debug;
 
 #[cfg(feature = "payouts")]
 use common_utils::request::RequestContent;
+use common_utils::types::{AmountConvertor, StringMajorUnit, StringMajorUnitForConnector};
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise.rs:5]:
```suggestion
use common_utils::types::{AmountConvertor, MinorUnit, MinorUnitForConnector};
```

File: crates/router/src/connector/wise/transformers.rs [hunk 4]
```diff
@@ -16,6 +17,20 @@ use crate::{
     },
 };
 use crate::{core::errors, types};
+#[derive(Debug, Serialize)]
+pub struct WiseRouterData<T> {
+    pub amount: StringMajorUnit,
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:22]:
```suggestion
    pub amount: MinorUnit,
```

File: crates/router/src/connector/wise/transformers.rs [hunk 5]
```diff
@@ -16,6 +17,20 @@ use crate::{
...
+    pub router_data: T,
+}
+
+impl<T> From<(StringMajorUnit, T)> for WiseRouterData<T> {
+    fn from((amount, router_data): (StringMajorUnit, T)) -> Self {
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:27]:
```suggestion
    fn from((amount, router_data): (MinorUnit, T)) -> Self {
```

File: crates/router/src/connector/wise/transformers.rs [hunk 6]
```diff
@@ -16,6 +17,20 @@ use crate::{
     },
 };
 use crate::{core::errors, types};
+#[derive(Debug, Serialize)]
+pub struct WiseRouterData<T> {
+    pub amount: StringMajorUnit,
+    pub router_data: T,
+}
+
+impl<T> From<(StringMajorUnit, T)> for WiseRouterData<T> {
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:26]:
```suggestion
impl<T> From<(MinorUnit, T)> for WiseRouterData<T> {
```

File: crates/router/src/connector/wise/transformers.rs [hunk 7]
```diff
@@ -61,7 +76,7 @@ pub struct SubError {
 #[derive(Debug, Serialize)]
 #[serde(rename_all = "camelCase")]
 pub struct WiseRecipientCreateRequest {
-    currency: String,
+    currency: StringMajorUnit,
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:79]: Please revert this change

File: crates/router/src/connector/wise/transformers.rs [hunk 8]
```diff
@@ -156,8 +171,8 @@ pub struct WiseRecipientCreateResponse {
 pub struct WisePayoutQuoteRequest {
     source_currency: String,
     target_currency: String,
-    source_amount: Option<i64>,
-    target_amount: Option<i64>,
+    source_amount: Option<StringMajorUnit>,
+    target_amount: Option<StringMajorUnit>,
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:175]:
```suggestion
    source_amount: Option<MinorUnit>,
    target_amount: Option<MinorUnit>,
```

File: crates/router/src/connector/wise/transformers.rs [hunk 9]
```diff
@@ -383,7 +401,7 @@ impl<F> TryFrom<&types::PayoutsRouterData<F>> for WiseRecipientCreateRequest {
                     })?;
                 Ok(Self {
                     profile: source_id,
-                    currency: request.destination_currency.to_string(),
+                    currency: item_data.amount.clone(),
```
Comment (deepanshu-iiitu) [inline crates/router/src/connector/wise/transformers.rs:404]: Please revert this change.

**Review by deepanshu-iiitu — CHANGES_REQUESTED** [review]

**Comment by deepanshu-iiitu** [comment]
Hey @shuklamaneesh23 
The amount in this payment processor should be `MinorUnit` and not `StringMajorUnit`. Please implement the above suggested changes.
Also currency should be of type string.

**Comment by shuklamaneesh23** [comment]
hey @deepanshu-iiitu , I have updated the code as per the suggested change.
Please let me know if I could fix something else.
Thanks!

**Review by deepanshu-iiitu — APPROVED** [review]

**Review by Sakilmostak — APPROVED** [review]

**Review by dracarys18 — APPROVED** [review]

**Pull request merged**